File: internal/analyzers/clones/analyzer.go
package clones

import (
	"encoding/json"
	"fmt"
	"io"
	"strings"

	"gopkg.in/yaml.v3"

	"github.com/Sumatoshi-tech/codefang/internal/analyzers/analyze"
	"github.com/Sumatoshi-tech/codefang/internal/analyzers/common"
	"github.com/Sumatoshi-tech/codefang/internal/analyzers/common/renderer"
	"github.com/Sumatoshi-tech/codefang/internal/analyzers/common/reportutil"
	"github.com/Sumatoshi-tech/codefang/internal/analyzers/common/terminal"
	"github.com/Sumatoshi-tech/codefang/pkg/alg/lsh"
	"github.com/Sumatoshi-tech/codefang/pkg/alg/minhash"
	"github.com/Sumatoshi-tech/codefang/pkg/pipeline"
	"github.com/Sumatoshi-tech/codefang/pkg/uast/pkg/node"
)

// Analysis configuration constants.
const (
	// numHashes is the number of MinHash hash functions per signature.
	numHashes = 128

	// numBands is the number of LSH bands.
	numBands = 16

	// numRows is the number of rows per LSH band.
	numRows = 8

	// minFunctionNodes is the minimum number of AST nodes a function must have
	// to be included in clone detection. Functions below this threshold are
	// trivial (getters, setters, return-nil stubs) and produce false positives
	// because their minimal AST structure hashes identically regardless of purpose.
	// Empirical: getters ≈ 13-15 nodes, setters ≈ 19, real logic ≥ 25.
	minFunctionNodes = 20

	// analyzerName is the registered name of the clone detection analyzer.
	analyzerName = "clones"

	// analyzerFlag is the CLI flag for the clone detection analyzer.
	analyzerFlag = "clone-detection"

	// analyzerDescription is the human-readable description.
	analyzerDescription = "Detects duplicate and near-duplicate code using MinHash and LSH."

	// analyzerID is the full analyzer ID for registration.
	analyzerID = "static/clones"
)

// Threshold constants for the Thresholds() method.
const (
	thresholdCloneRatioYellow = 0.1
	thresholdCloneRatioRed    = 0.3
	thresholdClonePairsYellow = 5
	thresholdClonePairsRed    = 20
)

// Message constants.
const (
	msgNoClones     = "No code clones detected"
	msgLowClones    = "Low duplication - few clone pairs detected"
	msgModClones    = "Moderate duplication - consider refactoring clone pairs"
	msgHighClones   = "High duplication - significant refactoring recommended"
	msgNoFunctions  = "No functions found for clone analysis"
	msgEmptyAST     = "No AST provided"
	msgAnalysisOK   = "Clone analysis completed"
	pairCountLow    = 5
	pairCountMod    = 15
	maxTraversalVal = 10
)

// Configuration option keys for the clones analyzer.
const (
	ConfigClonesMaxClonePairs        = "Clones.MaxClonePairs"
	ConfigClonesNumHashes            = "Clones.NumHashes"
	ConfigClonesNumBands             = "Clones.NumBands"
	ConfigClonesNumRows              = "Clones.NumRows"
	ConfigClonesShingleSize          = "Clones.ShingleSize"
	ConfigClonesSimilarityType2      = "Clones.SimilarityType2"
	ConfigClonesSimilarityType3      = "Clones.SimilarityType3"
	ConfigClonesThresholdRatioYellow = "Clones.ThresholdRatioYellow"
	ConfigClonesThresholdRatioRed    = "Clones.ThresholdRatioRed"
	ConfigClonesThresholdPairsYellow = "Clones.ThresholdPairsYellow"
	ConfigClonesThresholdPairsRed    = "Clones.ThresholdPairsRed"
)

// Analyzer provides clone detection analysis using MinHash and LSH.
type Analyzer struct {
	traverser *common.UASTTraverser
	shingler  *Shingler

	cfgMaxClonePairs        int
	cfgNumHashes            int
	cfgNumBands             int
	cfgNumRows              int
	cfgSimilarityType2      float64
	cfgSimilarityType3      float64
	cfgThresholdRatioYellow float64
	cfgThresholdRatioRed    float64
	cfgThresholdPairsYellow int
	cfgThresholdPairsRed    int
}

// NewAnalyzer creates a new clone detection Analyzer.
func NewAnalyzer() *Analyzer {
	return &Analyzer{
		traverser: common.NewUASTTraverser(common.TraversalConfig{
			MaxDepth:    maxTraversalVal,
			IncludeRoot: true,
		}),
		shingler:                NewShingler(defaultShingleSize),
		cfgNumHashes:            numHashes,
		cfgNumBands:             numBands,
		cfgNumRows:              numRows,
		cfgSimilarityType2:      similarityType2,
		cfgSimilarityType3:      similarityType3,
		cfgThresholdRatioYellow: thresholdCloneRatioYellow,
		cfgThresholdRatioRed:    thresholdCloneRatioRed,
		cfgThresholdPairsYellow: thresholdClonePairsYellow,
		cfgThresholdPairsRed:    thresholdClonePairsRed,
	}
}

// Name returns the analyzer name.
func (a *Analyzer) Name() string {
	return analyzerName
}

// Flag returns the CLI flag for the analyzer.
func (a *Analyzer) Flag() string {
	return analyzerFlag
}

// Descriptor returns stable analyzer metadata.
func (a *Analyzer) Descriptor() analyze.Descriptor {
	return analyze.NewDescriptor(
		analyze.ModeStatic,
		a.Name(),
		analyzerDescription,
	)
}

// ListConfigurationOptions returns configuration options.
func (a *Analyzer) ListConfigurationOptions() []pipeline.ConfigurationOption {
	return []pipeline.ConfigurationOption{}
}

// Configure configures the analyzer.
func (a *Analyzer) Configure(facts map[string]any) error {
	if val, ok := facts[ConfigClonesMaxClonePairs].(int); ok {
		a.cfgMaxClonePairs = val
	}

	if val, ok := facts[ConfigClonesNumHashes].(int); ok {
		a.cfgNumHashes = val
	}

	if val, ok := facts[ConfigClonesNumBands].(int); ok {
		a.cfgNumBands = val
	}

	if val, ok := facts[ConfigClonesNumRows].(int); ok {
		a.cfgNumRows = val
	}

	if val, ok := facts[ConfigClonesShingleSize].(int); ok {
		a.shingler = NewShingler(val)
	}

	if val, ok := facts[ConfigClonesSimilarityType2].(float64); ok {
		a.cfgSimilarityType2 = val
	}

	if val, ok := facts[ConfigClonesSimilarityType3].(float64); ok {
		a.cfgSimilarityType3 = val
	}

	if val, ok := facts[ConfigClonesThresholdRatioYellow].(float64); ok {
		a.cfgThresholdRatioYellow = val
	}

	if val, ok := facts[ConfigClonesThresholdRatioRed].(float64); ok {
		a.cfgThresholdRatioRed = val
	}

	if val, ok := facts[ConfigClonesThresholdPairsYellow].(int); ok {
		a.cfgThresholdPairsYellow = val
	}

	if val, ok := facts[ConfigClonesThresholdPairsRed].(int); ok {
		a.cfgThresholdPairsRed = val
	}

	return nil
}

// Thresholds returns the color-coded thresholds for clone metrics.
func (a *Analyzer) Thresholds() analyze.Thresholds {
	return analyze.Thresholds{
		"clone_ratio": {
			"green":  0.0,
			"yellow": a.cfgThresholdRatioYellow,
			"red":    a.cfgThresholdRatioRed,
		},
		"total_clone_pairs": {
			"green":  0,
			"yellow": a.cfgThresholdPairsYellow,
			"red":    a.cfgThresholdPairsRed,
		},
	}
}

// CreateAggregator returns a new aggregator for clone analysis.
func (a *Analyzer) CreateAggregator() analyze.ResultAggregator {
	agg := NewAggregator()
	if a.cfgMaxClonePairs > 0 {
		agg.MaxClonePairs = a.cfgMaxClonePairs
	}

	agg.NumBands = a.cfgNumBands
	agg.NumRows = a.cfgNumRows
	agg.SimilarityType3 = a.cfgSimilarityType3

	return agg
}

// CreateVisitor creates a new visitor for single-pass traversal optimization.
func (a *Analyzer) CreateVisitor() analyze.AnalysisVisitor {
	v := NewVisitor()
	v.numHashes = a.cfgNumHashes
	v.shingler = a.shingler
	v.similarityType2 = a.cfgSimilarityType2
	v.similarityType3 = a.cfgSimilarityType3

	return v
}

// CreateReportSection creates a ReportSection from report data.
func (a *Analyzer) CreateReportSection(report analyze.Report) analyze.ReportSection {
	return NewReportSection(report)
}

// Analyze performs clone detection on the given UAST.
func (a *Analyzer) Analyze(root *node.Node) (analyze.Report, error) {
	if root == nil {
		return buildEmptyReport(msgEmptyAST), nil
	}

	functions := a.findFunctions(root)
	if len(functions) == 0 {
		return buildEmptyReport(msgNoFunctions), nil
	}

	pairs := a.detectClones(functions)

	return a.buildReport(len(functions), pairs), nil
}

// findFunctions finds all function and method nodes in the UAST.
func (a *Analyzer) findFunctions(root *node.Node) []*node.Node {
	typeNodes := a.traverser.FindNodesByType(root, []string{node.UASTFunction, node.UASTMethod})
	roleNodes := a.traverser.FindNodesByRoles(root, []string{node.RoleFunction})

	seen := make(map[*node.Node]bool)

	var functions []*node.Node

	for _, n := range typeNodes {
		if !seen[n] && isFunctionNode(n) {
			seen[n] = true

			functions = append(functions, n)
		}
	}

	for _, n := range roleNodes {
		if !seen[n] && isFunctionNode(n) {
			seen[n] = true

			functions = append(functions, n)
		}
	}

	return functions
}

// isFunctionNode checks if a node represents a function.
func isFunctionNode(n *node.Node) bool {
	if n == nil {
		return false
	}

	return n.HasAnyType(node.UASTFunction, node.UASTMethod) ||
		n.HasAllRoles(node.RoleFunction, node.RoleDeclaration)
}

// funcEntry holds a function's name and MinHash signature for clone detection.
type funcEntry struct {
	name string
	sig  *minhash.Signature
}

// detectClones builds MinHash signatures and uses LSH to find clone pairs.
func (a *Analyzer) detectClones(functions []*node.Node) []ClonePair {
	entries := a.buildSignatures(functions)
	if len(entries) == 0 {
		return nil
	}

	idx, err := lsh.New(a.cfgNumBands, a.cfgNumRows)
	if err != nil {
		return nil
	}

	for _, entry := range entries {
		insertErr := idx.Insert(entry.name, entry.sig)
		if insertErr != nil {
			continue
		}
	}

	// Per-file detection: no cap (single-file scope, bounded by function count).
	result := findClonePairs(entries, idx, 0, a.cfgSimilarityType3)

	return result.pairs
}

// buildSignatures computes MinHash signatures for all functions.
func (a *Analyzer) buildSignatures(functions []*node.Node) []funcEntry {
	entries := make([]funcEntry, 0, len(functions))

	for _, fn := range functions {
		if countNodes(fn) < minFunctionNodes {
			continue
		}

		shingles := a.shingler.ExtractShingles(fn)
		if len(shingles) == 0 {
			continue
		}

		sig, err := minhash.New(a.cfgNumHashes)
		if err != nil {
			continue
		}

		for _, shingle := range shingles {
			sig.Add(shingle)
		}

		name := extractFuncName(fn)

		entries = append(entries, funcEntry{
			name: name,
			sig:  sig,
		})
	}

	return entries
}

// extractFuncName extracts a unique function name from a node.
// For methods, qualifies with the receiver type (e.g., "Foo.DoWork") to avoid
// collisions in the LSH index when different types share the same method name.
func extractFuncName(fn *node.Node) string {
	name, ok := common.ExtractEntityName(fn)
	if !ok || name == "" {
		if fn.Token != "" {
			name = fn.Token
		} else {
			name = string(fn.Type)
		}
	}

	if fn.Type == node.UASTMethod {
		if recv := extractReceiverType(fn); recv != "" {
			return recv + "." + name
		}
	}

	return name
}

// extractReceiverType extracts the receiver type name from a Method node.
// The UAST represents the receiver as the first Parameter child with a token
// like "(f *Foo)" or "(f Foo)".
func extractReceiverType(fn *node.Node) string {
	for _, child := range fn.Children {
		if !child.HasAnyRole(node.RoleParameter) {
			continue
		}

		// The receiver parameter token contains the full "(name *Type)" text.
		tok := child.Token
		if tok == "" {
			continue
		}

		// Extract the type name: strip parens, pointer star, and variable name.
		// Strip parens, pointer star, and variable name to extract the type.
		tok = strings.TrimPrefix(tok, "(")
		tok = strings.TrimSuffix(tok, ")")
		tok = strings.TrimSpace(tok)

		// Split "f *Foo" into parts, take the last one (the type).
		parts := strings.Fields(tok)
		// Receiver has at least two parts: variable name and type.
		const minReceiverParts = 2
		if len(parts) < minReceiverParts {
			continue
		}

		typeName := parts[len(parts)-1]
		typeName = strings.TrimPrefix(typeName, "*")

		if typeName != "" {
			return typeName
		}
	}

	return ""
}

// countNodes returns the total number of nodes in a subtree.
func countNodes(n *node.Node) int {
	if n == nil {
		return 0
	}

	count := 1

	for _, child := range n.Children {
		count += countNodes(child)
	}

	return count
}

// buildReport constructs the analysis report.
func (a *Analyzer) buildReport(totalFunctions int, pairs []ClonePair) analyze.Report {
	cloneRatio := computeCloneRatio(countDistinctFuncs(pairs), totalFunctions)
	message := cloneMessage(len(pairs))

	pairsForReport := make([]map[string]any, 0, len(pairs))

	for _, p := range pairs {
		pairsForReport = append(pairsForReport, map[string]any{
			"func_a":     p.FuncA,
			"func_b":     p.FuncB,
			"similarity": p.Similarity,
			"clone_type": p.CloneType,
		})
	}

	return analyze.Report{
		keyAnalyzerName:    analyzerName,
		keyTotalFunctions:  totalFunctions,
		keyTotalClonePairs: len(pairs),
		keyCloneRatio:      cloneRatio,
		keyClonePairs:      pairsForReport,
		keyMessage:         message,
	}
}

// buildEmptyReport creates an empty report with the given message.
func buildEmptyReport(message string) analyze.Report {
	return common.NewResultBuilder().BuildCustomEmptyResult(map[string]any{
		keyAnalyzerName:    analyzerName,
		keyTotalFunctions:  0,
		keyTotalClonePairs: 0,
		keyCloneRatio:      0.0,
		keyMessage:         message,
	})
}

// countDistinctFuncs returns the number of unique function names across all pairs.
func countDistinctFuncs(pairs []ClonePair) int {
	unique := make(map[string]struct{}, len(pairs))

	for idx := range pairs {
		unique[pairs[idx].FuncA] = struct{}{}
		unique[pairs[idx].FuncB] = struct{}{}
	}

	return len(unique)
}

// computeCloneRatio calculates the fraction of functions involved in at least one clone pair.
// Returns a value in [0, 1]: 0 means no duplication, 1 means every function has a clone.
func computeCloneRatio(clonedFuncs, totalFunctions int) float64 {
	if totalFunctions == 0 || clonedFuncs == 0 {
		return 0.0
	}

	return float64(clonedFuncs) / float64(totalFunctions)
}

// cloneMessage returns a human-readable message based on clone pair count.
func cloneMessage(pairCount int) string {
	if pairCount == 0 {
		return msgNoClones
	}

	if pairCount <= pairCountLow {
		return msgLowClones
	}

	if pairCount <= pairCountMod {
		return msgModClones
	}

	return msgHighClones
}

// FormatReport formats clone analysis results as human-readable text.
func (a *Analyzer) FormatReport(report analyze.Report, w io.Writer) error {
	section := NewReportSection(report)
	config := terminal.NewConfig()
	r := renderer.NewSectionRenderer(config.Width, false, config.NoColor)

	_, err := fmt.Fprint(w, r.Render(section))
	if err != nil {
		return fmt.Errorf("formatreport: %w", err)
	}

	return nil
}

// FormatReportJSON formats clone analysis results as JSON.
func (a *Analyzer) FormatReportJSON(report analyze.Report, w io.Writer) error {
	metrics := computeMetricsFromReport(report)

	jsonData, err := json.MarshalIndent(metrics, "", "  ")
	if err != nil {
		return fmt.Errorf("formatreportjson: %w", err)
	}

	_, err = fmt.Fprint(w, string(jsonData))
	if err != nil {
		return fmt.Errorf("formatreportjson: %w", err)
	}

	return nil
}

// FormatReportYAML formats clone analysis results as YAML.
func (a *Analyzer) FormatReportYAML(report analyze.Report, w io.Writer) error {
	metrics := computeMetricsFromReport(report)

	data, err := yaml.Marshal(metrics)
	if err != nil {
		return fmt.Errorf("formatreportyaml: %w", err)
	}

	_, err = w.Write(data)
	if err != nil {
		return fmt.Errorf("formatreportyaml: %w", err)
	}

	return nil
}

// FormatReportBinary formats clone analysis results as binary envelope.
func (a *Analyzer) FormatReportBinary(report analyze.Report, w io.Writer) error {
	metrics := computeMetricsFromReport(report)

	err := reportutil.EncodeBinaryEnvelope(metrics, w)
	if err != nil {
		return fmt.Errorf("formatreportbinary: %w", err)
	}

	return nil
}

// FormatReportPlot formats clone analysis results as HTML plot.
func (a *Analyzer) FormatReportPlot(report analyze.Report, w io.Writer) error {
	sections, err := a.generatePlotSections(report)
	if err != nil {
		return fmt.Errorf("formatreportplot: %w", err)
	}

	return renderPlotSections(sections, w)
}

File: rust/crates/cf-clones/src/analyzer.rs
//! The per-file clone-detection [`Analyzer`].
//!
//! Port of `internal/analyzers/clones/analyzer.go`. It builds MinHash signatures
//! for every function in a UAST, indexes them with LSH, finds the clone pairs,
//! and produces the `analyze::Report` map plus the machine-format projections.

use std::io::Write;

use cf_analyze::descriptor::{new_descriptor, AnalyzerMode, Descriptor};
use cf_analyze::history::AnalyzerError;
use cf_analyze::{GoMap, GoValue, MapOrigin, Report, Thresholds};
use cf_uast_node::Node;

use crate::engine::{
    build_index, build_signatures, count_distinct_funcs, find_clone_pairs, is_function_node,
};
use crate::uast::{ROLE_FUNCTION, UAST_FUNCTION, UAST_METHOD};
use crate::report::{
    categorize_clone_pairs, clone_type_dist_map, ClonePair, ComputedMetrics, SIMILARITY_TYPE3,
};
use crate::shingler::Shingler;
use crate::{
    clone_message, compute_clone_ratio, ANALYZER_DESCRIPTION, ANALYZER_FLAG, ANALYZER_NAME,
    KEY_ANALYZER_NAME, KEY_CLONE_PAIRS, KEY_CLONE_RATIO, KEY_CLONE_TYPE_DISTRIBUTION, KEY_MESSAGE,
    KEY_TOTAL_CLONE_PAIRS, KEY_TOTAL_FUNCTIONS, MSG_EMPTY_AST, MSG_NO_FUNCTIONS, NUM_BANDS,
    NUM_HASHES, NUM_ROWS, THRESHOLD_CLONE_PAIRS_RED, THRESHOLD_CLONE_PAIRS_YELLOW,
    THRESHOLD_CLONE_RATIO_RED, THRESHOLD_CLONE_RATIO_YELLOW,
};

/// Clone-detection analyzer using MinHash and LSH. Mirrors Go `Analyzer`.
#[derive(Debug, Clone)]
pub struct Analyzer {
    shingler: Shingler,
    cfg_num_hashes: usize,
    cfg_num_bands: usize,
    cfg_num_rows: usize,
    cfg_similarity_type3: f64,
}

impl Default for Analyzer {
    fn default() -> Self {
        Self::new()
    }
}

impl Analyzer {
    /// Creates a new clone-detection analyzer with the Go defaults. Mirrors Go
    /// `NewAnalyzer`.
    #[must_use]
    pub fn new() -> Self {
        Self {
            shingler: Shingler::default(),
            cfg_num_hashes: NUM_HASHES,
            cfg_num_bands: NUM_BANDS,
            cfg_num_rows: NUM_ROWS,
            cfg_similarity_type3: SIMILARITY_TYPE3,
        }
    }

    /// The analyzer name. Mirrors Go `Analyzer.Name`.
    #[must_use]
    pub fn name(&self) -> &'static str {
        ANALYZER_NAME
    }

    /// The CLI flag. Mirrors Go `Analyzer.Flag`.
    #[must_use]
    pub fn flag(&self) -> &'static str {
        ANALYZER_FLAG
    }

    /// Stable analyzer metadata. Mirrors Go `Analyzer.Descriptor`.
    #[must_use]
    pub fn descriptor(&self) -> Descriptor {
        new_descriptor(AnalyzerMode::Static, ANALYZER_NAME, ANALYZER_DESCRIPTION)
    }

    /// Color-coded thresholds for clone metrics. Mirrors Go `Analyzer.Thresholds`.
    #[must_use]
    pub fn thresholds(&self) -> Thresholds {
        let mut ratio = GoMap::new(MapOrigin::Map);
        ratio.push("green", GoValue::Float(0.0));
        ratio.push("yellow", GoValue::Float(THRESHOLD_CLONE_RATIO_YELLOW));
        ratio.push("red", GoValue::Float(THRESHOLD_CLONE_RATIO_RED));

        let mut pairs = GoMap::new(MapOrigin::Map);
        pairs.push("green", GoValue::Int(0));
        pairs.push("yellow", GoValue::Int(THRESHOLD_CLONE_PAIRS_YELLOW));
        pairs.push("red", GoValue::Int(THRESHOLD_CLONE_PAIRS_RED));

        let mut t = GoMap::new(MapOrigin::Map);
        t.push("clone_ratio", GoValue::Object(ratio));
        t.push("total_clone_pairs", GoValue::Object(pairs));
        t
    }

    /// Analyzes a UAST root into a clone-detection [`Report`].
    ///
    /// Mirrors Go `Analyzer.Analyze`. A `None` root yields the "No AST provided"
    /// empty report; no functions yields the "No functions found" empty report.
    #[must_use]
    pub fn analyze_node(&self, root: Option<&Node>) -> Report {
        let Some(root) = root else {
            return build_empty_report(MSG_EMPTY_AST);
        };

        let functions = self.find_functions(root);
        if functions.is_empty() {
            return build_empty_report(MSG_NO_FUNCTIONS);
        }

        let pairs = self.detect_clones(&functions);
        self.build_report(functions.len(), &pairs)
    }

    /// Finds all function and method nodes in the UAST.
    ///
    /// Mirrors Go `findFunctions`: collects by type (`Function`/`Method`) then by
    /// the `Function` role, deduplicating by node identity in that order.
    fn find_functions<'a>(&self, root: &'a Node) -> Vec<&'a Node> {
        let type_nodes = root.find(|n| n.has_any_type(&[UAST_FUNCTION, UAST_METHOD]));
        let role_nodes = root.find(|n| n.has_any_role(&[ROLE_FUNCTION]));

        let mut seen: Vec<*const Node> = Vec::new();
        let mut functions: Vec<&Node> = Vec::new();

        let consider = |n: &'a Node, seen: &mut Vec<*const Node>, out: &mut Vec<&'a Node>| {
            let ptr = std::ptr::from_ref::<Node>(n);
            if !seen.contains(&ptr) && is_function_node(n) {
                seen.push(ptr);
                out.push(n);
            }
        };

        for n in type_nodes {
            consider(n, &mut seen, &mut functions);
        }
        for n in role_nodes {
            consider(n, &mut seen, &mut functions);
        }

        functions
    }

    /// Builds MinHash signatures, indexes them, and finds clone pairs.
    ///
    /// Mirrors Go `detectClones` (per-file: no cap, threshold = Type-3).
    fn detect_clones(&self, functions: &[&Node]) -> Vec<ClonePair> {
        let entries = build_signatures(functions, &self.shingler, self.cfg_num_hashes);
        if entries.is_empty() {
            return Vec::new();
        }

        let Some(idx) = build_index(&entries, self.cfg_num_bands, self.cfg_num_rows) else {
            return Vec::new();
        };

        // Per-file detection: no cap (single-file scope, bounded by func count).
        find_clone_pairs(&entries, &idx, 0, self.cfg_similarity_type3).pairs
    }

    /// Constructs the analysis report map. Mirrors Go `buildReport`.
    fn build_report(&self, total_functions: usize, pairs: &[ClonePair]) -> Report {
        let clone_ratio = compute_clone_ratio(count_distinct_funcs(pairs), total_functions);
        let message = clone_message(pairs.len());

        let pairs_value =
            GoValue::Array(pairs.iter().map(ClonePair::to_go_value).collect::<Vec<_>>());
        let dist = clone_type_dist_map(categorize_clone_pairs(pairs));

        // analyze::Report is a map-origin GoMap: keys byte-sort on encode.
        let mut report = GoMap::new(MapOrigin::Map);
        report.push(KEY_ANALYZER_NAME, GoValue::Str(ANALYZER_NAME.to_string()));
        report.push(KEY_TOTAL_FUNCTIONS, GoValue::Int(total_functions as i64));
        report.push(KEY_TOTAL_CLONE_PAIRS, GoValue::Int(pairs.len() as i64));
        report.push(KEY_CLONE_RATIO, GoValue::Float(clone_ratio));
        report.push(KEY_CLONE_PAIRS, pairs_value);
        report.push(KEY_CLONE_TYPE_DISTRIBUTION, dist);
        report.push(KEY_MESSAGE, GoValue::Str(message.to_string()));
        report
    }

    /// Writes the report as indented JSON of [`ComputedMetrics`].
    ///
    /// Mirrors Go `FormatReportJSON` (`json.MarshalIndent("", "  ")`). Routes
    /// through [`cf_gojson`], never `serde_json`.
    ///
    /// # Errors
    /// Returns [`AnalyzerError`] if the writer fails.
    pub fn format_report_json(&self, report: &Report, w: &mut dyn Write) -> Result<(), AnalyzerError> {
        let metrics = compute_metrics_from_report(report);
        let bytes = cf_gojson::marshal_indent(&metrics.to_go_value());
        w.write_all(&bytes)
            .map_err(|e| AnalyzerError::Other(format!("formatreportjson: {e}")))
    }

    /// Writes the report as a CFB1 binary envelope of [`ComputedMetrics`].
    ///
    /// Mirrors Go `FormatReportBinary` (`reportutil.EncodeBinaryEnvelope`).
    ///
    /// # Errors
    /// Returns [`AnalyzerError`] if encoding or the writer fails.
    pub fn format_report_binary(
        &self,
        report: &Report,
        w: &mut dyn Write,
    ) -> Result<(), AnalyzerError> {
        let metrics = compute_metrics_from_report(report);
        let bytes = cf_reportutil::encode_binary_envelope(&metrics.to_go_value())
            .map_err(|e| AnalyzerError::Other(format!("formatreportbinary: {e}")))?;
        w.write_all(&bytes)
            .map_err(|e| AnalyzerError::Other(format!("formatreportbinary: {e}")))
    }

    /// Returns the [`ComputedMetrics`] projection used by the YAML path.
    ///
    /// Mirrors the `metrics := computeMetricsFromReport(report)` step shared by
    /// Go's `FormatReportYAML`. The YAML emitter (`cf-goyaml`) is wired by the
    /// framework once it lands; YAML is a non-binding capture (DESIGN §6), so the
    /// projection — not the emitter — is the byte-identity-critical part and is
    /// exposed here for it.
    #[must_use]
    pub fn computed_metrics(&self, report: &Report) -> ComputedMetrics {
        compute_metrics_from_report(report)
    }
}

/// Creates an empty report with the given message. Mirrors Go `buildEmptyReport`.
///
/// The Go path runs through `common.NewResultBuilder().BuildCustomEmptyResult`,
/// which returns exactly these five keys as a `map[string]any`.
#[must_use]
pub fn build_empty_report(message: &str) -> Report {
    let mut report = GoMap::new(MapOrigin::Map);
    report.push(KEY_ANALYZER_NAME, GoValue::Str(ANALYZER_NAME.to_string()));
    report.push(KEY_TOTAL_FUNCTIONS, GoValue::Int(0));
    report.push(KEY_TOTAL_CLONE_PAIRS, GoValue::Int(0));
    report.push(KEY_CLONE_RATIO, GoValue::Float(0.0));
    report.push(KEY_MESSAGE, GoValue::Str(message.to_string()));
    report
}

/// Projects a report map into [`ComputedMetrics`]. Mirrors Go
/// `computeMetricsFromReport`.
#[must_use]
pub fn compute_metrics_from_report(report: &Report) -> ComputedMetrics {
    let total_functions = cf_reportutil::get_int(report, KEY_TOTAL_FUNCTIONS);
    let total_clone_pairs = cf_reportutil::get_int(report, KEY_TOTAL_CLONE_PAIRS);
    let clone_ratio = cf_reportutil::get_float64(report, KEY_CLONE_RATIO);
    let message = cf_reportutil::get_string(report, KEY_MESSAGE);
    let clone_pairs = extract_clone_pairs(report);
    let clone_type_dist = extract_clone_type_dist(report);

    ComputedMetrics {
        total_functions,
        total_clone_pairs,
        clone_ratio,
        clone_type_dist,
        clone_pairs,
        message,
    }
}

/// Extracts the clone pairs from a report. Mirrors Go `extractClonePairs`
/// (handling the `[]map[string]any` case produced by [`Analyzer::build_report`]).
fn extract_clone_pairs(report: &Report) -> Vec<ClonePair> {
    let Some(GoValue::Array(items)) = cf_reportutil::get(report, KEY_CLONE_PAIRS) else {
        return Vec::new();
    };

    let mut pairs = Vec::with_capacity(items.len());
    for item in items {
        if let GoValue::Map(m) = item {
            pairs.push(clone_pair_from_map(m));
        }
    }
    pairs
}

/// Extracts a [`ClonePair`] from a map value. Mirrors Go `clonePairFromMap`.
fn clone_pair_from_map(m: &GoMap) -> ClonePair {
    let mut pair = ClonePair {
        func_a: String::new(),
        func_b: String::new(),
        similarity: 0.0,
        clone_type: String::new(),
    };
    for (k, v) in m.entries() {
        match (k.as_str(), v) {
            ("func_a", GoValue::Str(s)) => pair.func_a = s.clone(),
            ("func_b", GoValue::Str(s)) => pair.func_b = s.clone(),
            ("similarity", GoValue::Float(f)) => pair.similarity = *f,
            ("clone_type", GoValue::Str(s)) => pair.clone_type = s.clone(),
            _ => {}
        }
    }
    pair
}

/// Extracts the `clone_type_distribution` counts, if present, mirroring the Go
/// `report[keyCloneTypeDistribution].(map[string]int)` branch of
/// `computeMetricsFromReport`.
fn extract_clone_type_dist(report: &Report) -> Option<crate::report::CloneTypeCounts> {
    let Some(GoValue::Map(m)) =
        cf_reportutil::get(report, KEY_CLONE_TYPE_DISTRIBUTION)
    else {
        return None;
    };

    let mut counts = crate::report::CloneTypeCounts::default();
    for (k, v) in m.entries() {
        if let GoValue::Int(n) = v {
            match k.as_str() {
                crate::report::CLONE_TYPE1 => counts.type1 = *n,
                crate::report::CLONE_TYPE2 => counts.type2 = *n,
                crate::report::CLONE_TYPE3 => counts.type3 = *n,
                _ => {}
            }
        }
    }
    Some(counts)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::uast::NodeBuilder;
    use cf_gojson::Encoder;
    use cf_uast_node::Node;

    /// Builds a function subtree with at least `MIN_FUNCTION_NODES` nodes and the
    /// given identifier name, so it survives the size gate and produces shingles.
    fn function(name: &str) -> Node {
        let name_node = NodeBuilder::new("Identifier").role("Name").token(name).build();
        let mut f = NodeBuilder::new("Function").role("Function").child(name_node).build();
        // Add a deterministic body of ≥20 nodes so countNodes >= 20 and there
        // are enough types (>= k=5) to shingle.
        let mut block = NodeBuilder::new("Block").build();
        for i in 0..24 {
            let kind = ["Identifier", "Call", "Literal", "Operator"][i % 4];
            block.add_child(NodeBuilder::new(kind).build());
        }
        f.add_child(block);
        f
    }

    #[test]
    fn descriptor_is_static_clones() {
        let a = Analyzer::new();
        assert_eq!(a.descriptor().id, "static/clones");
        assert_eq!(a.name(), "clones");
        assert_eq!(a.flag(), "clone-detection");
    }

    #[test]
    fn nil_root_returns_empty_ast_message() {
        let a = Analyzer::new();
        let report = a.analyze_node(None);
        assert_eq!(cf_reportutil::get_string(&report, KEY_MESSAGE), MSG_EMPTY_AST);
        assert_eq!(cf_reportutil::get_int(&report, KEY_TOTAL_FUNCTIONS), 0);
    }

    #[test]
    fn no_functions_returns_no_functions_message() {
        let a = Analyzer::new();
        let root = NodeBuilder::new("File").build(); // no function nodes
        let report = a.analyze_node(Some(&root));
        assert_eq!(
            cf_reportutil::get_string(&report, KEY_MESSAGE),
            MSG_NO_FUNCTIONS
        );
    }

    #[test]
    fn identical_functions_detected_as_type1_clone() {
        let a = Analyzer::new();
        // Two structurally identical functions -> similarity 1.0 -> Type-1.
        let root = NodeBuilder::new("File")
            .child(function("foo"))
            .child(function("bar"))
            .build();
        let report = a.analyze_node(Some(&root));
        assert_eq!(cf_reportutil::get_int(&report, KEY_TOTAL_FUNCTIONS), 2);
        assert_eq!(cf_reportutil::get_int(&report, KEY_TOTAL_CLONE_PAIRS), 1);

        // clone_ratio = 2 distinct cloned funcs / 2 total = 1.0.
        assert!((cf_reportutil::get_float64(&report, KEY_CLONE_RATIO) - 1.0).abs() < 1e-12);

        let metrics = compute_metrics_from_report(&report);
        assert_eq!(metrics.clone_pairs.len(), 1);
        assert_eq!(metrics.clone_pairs[0].clone_type, "Type-1");
        assert!((metrics.clone_pairs[0].similarity - 1.0).abs() < 1e-9);
    }

    #[test]
    fn empty_report_json_round_trips_through_computed_metrics() {
        let a = Analyzer::new();
        let report = a.analyze_node(None);
        let mut buf = Vec::new();
        a.format_report_json(&report, &mut buf).expect("json ok");
        let json = String::from_utf8(buf).unwrap();
        // Indented; ComputedMetrics field order; no clone_type_distribution
        // (omitempty); clone_pairs null.
        assert!(json.contains("\"total_functions\": 0"));
        assert!(json.contains("\"clone_pairs\": null"));
        assert!(json.contains("\"message\": \"No AST provided\""));
        assert!(!json.contains("clone_type_distribution"));
    }

    #[test]
    fn thresholds_structure_matches_go() {
        let a = Analyzer::new();
        let json = Encoder::marshal().encode_to_string(&GoValue::Object(a.thresholds()));
        assert_eq!(
            json,
            concat!(
                r#"{"clone_ratio":{"green":0,"red":0.3,"yellow":0.1},"#,
                r#""total_clone_pairs":{"green":0,"red":20,"yellow":5}}"#
            )
        );
    }
}
